### PR TITLE
Refactored signature generation

### PR DESCRIPTION
**Pull request by osmirnov**
Last week I merged NeoEntityBase and WithHash256 and got Frankenstein. Mostly I did not like two things:

1.  Hash property in NeoEntityBase because it means different things for different inherited classes (it is obvious if we check JsonProperty name: JsonProperty("txid") makes no sense in BlockHeader class. In the original project it was serialized as "hash"). The solution was to push down Hash property in the class hierarchy and see over time how it goes.

2. How we calculated signature in WithHash256. It required two dependencies that were not in place (ICrypto, ISerializer). The reason is we do not inject dependencies into Domain Objects for the best. By the way, if you need this it immediately adds smell to the code. Nevertheless, I've checked the previous implementation and found out it is pretty the same meaning we inherited not only bad design but also potential bugs. What're the bugs? We get hash only first time when we querying Hash property, the consequent times we retrieve cached version. Now asked yourselves what happens if we change any field used into a hash calculation after we get our cached version. You will be right if you say the hash will be invalid. The solution here is to delegate hash calculation to a service class that will be called when we need to get the snapshot of hash based on current state of fields. Right now I am sure we need to recalculate hash before adding/updating entity in a repository (context). So I move this code to there.

A couple of questions from me:
1. Why do we need DataEntryPrefix in IDbModel? We are going to store entities in different tables. If we have storage does not support containers then hash prefixing has to be hidden in DbModel implementation.

2. Why do we need Add methods in TransactionContext and BlockHeaderContext? The blockchain is aggregating transactions into blocks and saves them. As the result, we need BlockContext that is splitting every Block into BlockHeader and Transaction tables. Yes, we can read after from either of them but there is no such a thing as "add transaction" outside of the block context.

### CODE REVIEW

**Comment by aboimpinto** [comment]
About the DataEntryPrefix, it's just a copy from old methods that had this prefix .. I have no idea if we can avoid that or not. 
Who is going to make the final implementation of the DB layer should check the necessity of usage of the prefix.

**Comment by aboimpinto** [comment]
for the 2º point, I mention that in my video. 
maybe we should create a high level class that will have blockHeaderContext and transactionContract as dependency where we can perform elaborated operations like get transaction from a Block or others.

the concept in place allow the creation of this high level objects.

**Comment by aboimpinto** [comment]
about the first 2 points, well, I just include a proposal from some one .. I don't remember who.

make all the sense  to have Hash property and not "txid". This field will be part of the BlockHeader and Transaction, because, both need to have Hash and we do not need to defined "blockid" or "txid"

right?

**Comment by osmirnov** [comment]
About "blockid" and "txid"... It's ok if we can go with "hash" only but I feel other tools will expect more clarity from api what hash means (in our case we use it for hash, for id, for signature, etc.)

File: src/NeoSharp.BinarySerialization/BinarySerializerSettings.cs [hunk 1]
```diff
@@ -5,8 +5,8 @@ namespace NeoSharp.BinarySerialization
     public class BinarySerializerSettings
     {
         /// <summary>
-        /// Filter by order
+        /// Filter
         /// </summary>
-        public Func<int, bool> Filter { get; set; }
+        public Predicate<string> Filter { get; set; }
```
Comment (shargon) [inline src/NeoSharp.BinarySerialization/BinarySerializerSettings.cs]: I like this change

File: src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs [hunk 2]
```diff
@@ -142,22 +142,9 @@ public BinarySerializerCache(Type type)
...
-
-            foreach (BinarySerializerCacheEntry e in _entries)
-            {
-                if (haveFilter && !settings.Filter.Invoke(e.Context.Order)) continue;
-
-                ret += e.Serializer.Serialize(serializer, bw, e.GetValue(obj));
-            }
-
-            return ret;
+            return _serializer?.Serialize(serializer, bw, obj, settings) ?? _entries
```
Comment (shargon) [inline src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs]: I change this part, removing `_serializer`  please take a look

File: src/NeoSharp.Core/Persistence/Entity.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using NeoSharp.BinarySerialization;
+using NeoSharp.Core.Types;
+using Newtonsoft.Json;
+
+namespace NeoSharp.Core.Persistence
+{
+    public class Entity
+    {
```
Comment (shargon) [inline src/NeoSharp.Core/Persistence/Entity.cs]: Entity256 and Entity160? or is better `IEntity` ?

**Pull request merged**